File: java/GatoSimple.java
/**Modifica la clase Gato vista anteriormente añadiendo el método apareaCon.
Este método debe comprobar que los gatos son de distinto sexo, tras lo
cual, genera un nuevo gato. Por ejemplo, sería válido algo como Gato cria
= garfield.apareaCon(lisa). En caso de que los gatos sean del mismo sexo,
el método debe generar la excepción ExcepcionApareamientoImposible. Crea un
programa desde el que se pruebe el método.

@author Rafael Infante
*/
package excepciones;

/**
 * GatoSimple.java Definición de la clase GatoSimple
 * 
 * @author Rafael Infante
 */
public class GatoSimple {
	// atributos /////////////////////////////
	String color, raza, sexo;
	int edad;
	double peso;

	// métodos ///////////////////////////////
	// constructor
	GatoSimple(String s) {
		this.sexo = s;
	}

	// getter
	String getSexo() {
		return this.sexo;
	}

	/**
	 * Hace que el gato maulle
	 */
	void maulla() {
		System.out.println("Miauuuu");
	}

	/**
	 * Hace que el gato ronronee
	 */
	void ronronea() {
		System.out.println("mrrrrrr");
	}

	/**
	 * Hace que el gato coma. A los gatos les gusta el pescado, si le damos otra
	 * comida la rechazará.
	 *
	 * Programación orientada a objetos 123
	 * 
	 * @param comida la comida que se le ofrece al gato
	 */
	void come(String comida) {
		if (comida.equals("pescado")) {
			System.out.println("Hmmmm, gracias");
		} else {
			System.out.println("Lo siento, yo solo como pescado");
		}
	}

	/**
	 * Pone a pelear dos gatos. Solo se van a pelear dos machos entre sí.
	 *
	 * @param contrincante es el gato contra el que pelear
	 */
	void peleaCon(GatoSimple contrincante) {
		if (this.sexo.equals("hembra")) {
			System.out.println("no me gusta pelear");
		} else {
			if (contrincante.getSexo().equals("hembra")) {
				System.out.println("no peleo contra gatitas");
			} else {
				System.out.println("ven aquí que te vas a enterar");
			}
		}
	}

	public GatoSimple apareaCon(GatoSimple pareja) throws ExceptionApareamientoImposible {
		String s = "";
		if (!this.sexo.equals(pareja.sexo)) {
			s = (int) (Math.random() * 2) == 0 ? "hembra" : "macho";
		}else {
			//Lanzamos excepción
			throw new ExceptionApareamientoImposible("No se puede aparear son del mismo sexo.");
		}
		return new GatoSimple(s);

	}

	@Override
	public String toString() {
		return "Gato [color=" + color + ", raza=" + raza + ", sexo=" + sexo + ", edad=" + edad + ", peso=" + peso
				+ "]";
	}
	
	
}// fin clase GatoSimple

File: python/GatoSimple.py
'''
Modifica la clase Gato vista anteriormente añadiendo el método apareaCon.
Este método debe comprobar que los gatos son de distinto sexo, tras lo
cual, genera un nuevo gato. Por ejemplo, sería válido algo como Gato cria
= garfield.apareaCon(lisa). En caso de que los gatos sean del mismo sexo,
el método debe generar la excepción ExcepcionApareamientoImposible. Crea un
programa desde el que se pruebe el método.

@author: Rafael Infante
'''

from random import randint


class GatoSimple:  # clase gato simple
    
    def __init__(self, s):  # constructor
        self.sexo = s
        self.color = ""
        self.raza = ""
        self.edad = 0
        self.peso = 0
    
    # getter
    def obtenersexo(self):
        return self.sexo
  
    '''
    metodo que hace que el gato maulle
    '''

    def maulla(self):
        print("Miauuu")
    
    '''
    metodo que hace que el gato ronronee
    '''

    def ronronea(self):
        print("mrrrrr")
  
    '''
    metodo que hace que el gato coma.
    A los gatos les gusta el pescado, si le damos otra comida
    la rechazará.
  
    '''  

    def come(self, comida):
        if comida == "pescado":
            print("hmmm gracias")
        else:
            print("Lo siento, yo solo como pescado")
  
    '''
    metodo que hace que se peleen los gatos
    @param: objeto tipo gato
    '''

    def peleaCon(self, contrincante):
        if (self.sexo == "hembra"):
            print("no me gusta pelear")
        else:
            if (contrincante.obtenersexo() == "hembra"):
                print("no peleo contra gatitas");
            else:
                print("ven aqui que te vas a enterar");

    '''
    metodo que hace que se apareen 2 gatos de distinto sexo.
    @param: objeto tipo gato
    @return: objeto tipo gato
    '''

    def apareaCon(self, pareja):
        s = ""
        # if (!this.sexo.equals(pareja.sexo)) {
        if self.sexo == pareja.sexo:
            #lanzar excepcion
            raise ValueError("No se pueden aparear son del mismo sexo.")
            #print("No se pueden aparear son del mismo sexo.")
            #raise apareamientoError()
        lista = ["macho", "hembra"]
        s = lista[randint(0, 1)]

        return GatoSimple(s)

    '''
    metodo toString
    '''

    def __str__(self):
        return "Gato [color=" + self.color + ", raza=" + self.raza + ", sexo=" + self.sexo + ", edad=" + str(self.edad) + ", peso=" + str(self.peso) + "]"

    '''
    garfield = GatoSimple("macho")
    tom = GatoSimple("macho")
    lisa = GatoSimple("hembra")
    maria = GatoSimple("hembra")
    lisa.peleaCon(maria)
    '''
